File: my_first_ros_rclpy_pkg/my_first_ros_rclpy_pkg/helloworld_publisher.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile
from std_msgs.msg import String


class HelloworldPublisher(Node):  # Node 상속

    def __init__(self):
        super().__init__('helloworld_publisher')  # super이면 node.. 자기 부모의 인스턴스의 생성자를 초기화 해주기
        qos_profile = QoSProfile(depth=10)  # 통신 상태가 원할하지 못한 상황 퍼블리시할 데이터를 버퍼에 10개까지 저장
        self.helloworld_publisher = self.create_publisher(String, 'helloworld', qos_profile)
        # 1초마다 지정한 콜백함수를 실행하라는 것으로 아래 코드와 같이 설정하면 1초마다 publish_helloworld_msg 함수를 실행
        self.timer = self.create_timer(1, self.publish_helloworld_msg)
        self.count = 0

    def publish_helloworld_msg(self):
        msg = String()
        msg.data = 'Hello World: {0}'.format(self.count)
        self.helloworld_publisher.publish(msg)
        self.get_logger().info('Published message: {0}'.format(msg.data))
        self.count += 1


def main(args=None):
    rclpy.init(args=args)
    node = HelloworldPublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info('Keyboard Interrupt (SIGINT)')
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: my_first_ros_rclcpp_pkg/src/helloworld_publisher.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;
// 추후에 500ms, 1s과 같이 시간을 가식성을 높인 문자로 표현하기 위하여 namespace를 사용할 수 있도록 선언하였다.
// namespace 이름으로 묶는 기능, 함수이름이 같아도 다른 namespace이면 상관없다.

class HelloworldPublisher : public rclcpp::Node
{
public:
  HelloworldPublisher()
  : Node("helloworld_publisher"), count_(0)    // 생성 및 초기화
  {
    auto qos_profile = rclcpp::QoS(rclcpp::KeepLast(10));  // auto: 타입형은 알아서 설정
    helloworld_publisher_ = this->create_publisher<std_msgs::msg::String>(
      "helloworld", qos_profile);
    timer_ = this->create_wall_timer(
      1s, std::bind(&HelloworldPublisher::publish_helloworld_msg, this));

    // lambda 표현
    // timer_ = this->create_wall_timer(
    //   1s,
    //   [this]() -> void
    //     {
    //       auto msg = std_msgs::msg::String();
    //       msg.data = "Hello World2: " + std::to_string(count_++);
    //       RCLCPP_INFO(this->get_logger(), "Published message: '%s'", msg.data.c_str());
    //       helloworld_publisher_->publish(msg);
    //     }
    //   );
  }

private:
  void publish_helloworld_msg()
  {
    auto msg = std_msgs::msg::String();
    msg.data = "Hello World: " + std::to_string(count_++);
    RCLCPP_INFO(this->get_logger(), "Published message: '%s'", msg.data.c_str());
    helloworld_publisher_->publish(msg);
  }
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr helloworld_publisher_;  // SharedPtr 스마트 포인터로서 메모리 해제를 자동으로 해준다.
  size_t count_;
};


int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<HelloworldPublisher>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
